salt/modules/state.py

```diff
@@ -106,18 +106,17 @@ def _set_retcode(ret, highstate=None):
 
 def _get_pillar_errors(kwargs, pillar=None):
     """
-    Checks all pillars (external and internal) for errors.
-    Return an error message, if anywhere or None.
+    Check pillar for errors.
+
+    If a pillar is passed, it will be checked. Otherwise, the in-memory pillar
+    will checked instead. Passing kwargs['force'] = True short cuts the check
+    and always returns None, indicating no errors.
 
     :param kwargs: dictionary of options
-    :param pillar: external pillar
-    :return: None or an error message
+    :param pillar: pillar
+    :return: None or a list of error messages
     """
-    return (
-        None
-        if kwargs.get("force")
-        else (pillar or {}).get("_errors", __pillar__.get("_errors")) or None
-    )
+    return None if kwargs.get("force") else (pillar or __pillar__).get("_errors")
```

Comment:
Do these statements do different things?  Sorry if I'm not seeing it, but I just ran the tests you wrote with both versions of this `return` statement and got the same result.

Reply:
Thanks for taking a look! Let me try to explain it a bit, I hope this is comprehensible.

The old code looked at `pillar.get("_errors")` first and, if there were no errors, proceeded with `__pillar__.get("_errors")`. It checked the presence of the `"_errors"` key to determine if it would look at `__pillar__` or not.

The new code first checks if it should check `pillar` or `__pillar__` for the presence of the `"_errors"` key. Only one of these two dicts will be checked. Here the presence of a non-empty `pillar` is used to determine where to look for `"_errors"`.

The problem this is fixing is that you can have `_errors` in `__pillar__` (e.g. from the first pillar caching that happens on minion startup) that leak into `state.apply`. `state.apply` is getting new pillar data from the master anyway (`pillar` in this function), therefore it does not make sense to look for errors in the unused and outdated `__pillar__`.

Regarding the tests: Looks like I need to double check that they work as expected.

Reply:
Ah!  That makes sense to me, thank you for explaining!  Are you able to replicate what I found with the tests?

Also, sorry it took so long for me to take a look at this.

Reply:
Sorry, I haven't had the time yet. I'll get to it this week and will let you know when I have the results.

Reply:
I found the problem with the unit tests: I missed the case where the old and new logic differ :see_no_evil:. Initially I wrote the integration test to check the new logic, when I did the unit test I missed that case. I now added another case, where `__pillar__` has `_errors` and the "fresh" `pillar` does not. That case shows that the old logic reported errors while the new logic does not.